cpp/oneapi/dal/algo/knn/backend/model_impl.hpp

```diff
@@ -18,35 +18,118 @@
 
 #include "oneapi/dal/algo/knn/common.hpp"
 #include "oneapi/dal/algo/knn/backend/model_interop.hpp"
+#include "oneapi/dal/backend/serialization.hpp"
 
 namespace oneapi::dal::knn {
 
+#define KNN_SERIALIZABLE(Task, ClassificationId, SearchId)             \
+    ONEDAL_SERIALIZABLE_MAP2(Task,                                     \
+                             (task::classification, ClassificationId), \
+                             (task::search, SearchId))
+
 template <typename Task>
 class detail::v1::model_impl : public base {
 public:
-    model_impl() : interop_(nullptr) {}
+    model_impl() {}
     model_impl(const model_impl&) = delete;
     model_impl& operator=(const model_impl&) = delete;
 
-    model_impl(backend::model_interop* interop) : interop_(interop) {}
+    virtual backend::model_interop* get_interop() = 0;
+};
+
+namespace detail {
+namespace v1 {
+
+template <typename Task>
+class brute_force_model_impl : public model_impl<Task>,
+                               public KNN_SERIALIZABLE(Task,
+                                                       knn_brute_force_classification_model_impl_id,
+                                                       knn_brute_force_search_model_impl_id) {
+public:
+    brute_force_model_impl() : interop_(nullptr) {}
+    brute_force_model_impl(const brute_force_model_impl&) = delete;
+    brute_force_model_impl& operator=(const brute_force_model_impl&) = delete;
+
+    brute_force_model_impl(const table& data, const table& labels)
+            : data_(data),
+              labels_(labels),
+              interop_(nullptr) {}
 
-    ~model_impl() {
+    ~brute_force_model_impl() {
         delete interop_;
         interop_ = nullptr;
     }
 
-    backend::model_interop* get_interop() {
+    backend::model_interop* get_interop() override {
         return interop_;
     }
 
+    void serialize(dal::detail::output_archive& ar) const override {
+        ar(data_, labels_);
+        dal::detail::serialize_polymorphic(interop_, ar);
+    }
+
+    void deserialize(dal::detail::input_archive& ar) override {
+        ar(data_, labels_);
+        interop_ = dal::detail::deserialize_polymorphic<backend::model_interop>(ar);
+    }
+
+    table data_;
```

Comment:
private?

Reply:
for similar svm model internal parameters are public, do we need to make them private and use setters for their set?

Reply:
variable with name: "var_" is private field (align with code style). in svm used public fields. see [here](https://github.com/oneapi-src/oneDAL/blob/master/cpp/oneapi/dal/algo/svm/backend/model_impl.hpp#L39)

Reply:
oh, right, it is unaligned with code style, I will remove underscore, is it correct way to fix it?

Reply:
naming changed